File: packages/python/replica/starfish_replica/validate.py
"""Replica config validation.

Cross-references the ``remotes`` map (collection name → :class:`RemoteConfig`)
against the server's :class:`SyncConfig` collections. These rules were
previously inline in ``starfish-server``'s config validator; they moved here
when the ``remote`` field left the core ``CollectionConfig``.
"""

from __future__ import annotations

import re

from starfish_server.config.schema import CollectionConfig, SyncConfig
from starfish_server.constants import ENCRYPTION_DELEGATED

from starfish_replica.config import RemoteConfig, WriteMode

MIME_JSON = "application/json"


def _is_binary_collection(allowed_mime_types: list[str]) -> bool:
    return MIME_JSON not in [m.lower() for m in allowed_mime_types]


def validate_replica_config(
    config: SyncConfig,
    remotes: dict[str, RemoteConfig],
) -> list[str]:
    """Validate the replica configuration. Returns error messages (empty = valid)."""
    errors: list[str] = []
    by_name: dict[str, CollectionConfig] = {c.name: c for c in config.collections}

    for name, remote in remotes.items():
        col = by_name.get(name)
        if col is None:
            errors.append(
                f'Collection "{name}": remote replication configured for an unknown root collection'
            )
            continue

        if col.append_only:
            errors.append(f'Collection "{name}": appendOnly cannot be used with remote replication')
        if _is_binary_collection(col.allowed_mime_types):
            errors.append(f'Collection "{name}": binary collections cannot have remote replication')
        if re.search(r"\{[^}]+\}", col.storage_path):
            errors.append(
                f'Collection "{name}": remote collections must have a static storagePath '
                f'with no template variables (found "{col.storage_path}")'
            )
        if col.push_only:
            errors.append(f'Collection "{name}": remote collections cannot be pushOnly')
        if col.bundle:
            errors.append(f'Collection "{name}": remote collections cannot be part of a bundle')
        if col.encryption == ENCRYPTION_DELEGATED:
            errors.append(
                f'Collection "{name}": remote collections cannot use "{col.encryption}" encryption '
                f'(server cannot replicate opaque client-encrypted blobs)'
            )
        if remote.write_mode in (WriteMode.PUSH_THROUGH, WriteMode.BIDIRECTIONAL):
            if not remote.push_path:
                errors.append(
                    f'Collection "{name}": write_mode "{remote.write_mode.value}" '
                    f'requires remote.push_path to be set'
                )

    return errors

File: packages/ts/replica/src/validate.ts
import type { SyncConfig, CollectionConfig } from "@drakkar.software/starfish-server"
import { ENCRYPTION_DELEGATED } from "@drakkar.software/starfish-server"
import type { RemoteConfig } from "./config.js"

const MIME_JSON = "application/json"

function isBinaryCollection(allowedMimeTypes: string[]): boolean {
  return !allowedMimeTypes.some((m) => m.toLowerCase() === MIME_JSON)
}

/**
 * Validate the replica configuration by cross-referencing the `remotes` map
 * (collection name → `RemoteConfig`) against the server's `SyncConfig`
 * collections. Returns error messages (empty = valid).
 *
 * The core server schema no longer carries a `remote` field, so these
 * cross-cutting rules — which were previously inline in
 * `starfish-server`'s config validator — now live with the plugin that owns
 * the replica config.
 */
export function validateReplicaConfig(
  config: SyncConfig,
  remotes: Record<string, RemoteConfig>,
): string[] {
  const errors: string[] = []
  const byName = new Map<string, CollectionConfig>(
    config.collections.map((c) => [c.name, c]),
  )

  for (const [name, remote] of Object.entries(remotes)) {
    const col = byName.get(name)
    if (!col) {
      errors.push(
        `Collection "${name}": remote replication configured for an unknown root collection`,
      )
      continue
    }

    if (col.appendOnly) {
      errors.push(`Collection "${name}": appendOnly cannot be used with remote replication`)
    }
    if (isBinaryCollection(col.allowedMimeTypes)) {
      errors.push(`Collection "${name}": binary collections cannot have remote replication`)
    }
    if (/\{[^}]+\}/.test(col.storagePath)) {
      errors.push(
        `Collection "${name}": remote collections must have a static storagePath with no template variables (found "${col.storagePath}")`,
      )
    }
    if (col.pushOnly) {
      errors.push(`Collection "${name}": remote collections cannot be pushOnly`)
    }
    if (col.bundle) {
      errors.push(`Collection "${name}": remote collections cannot be part of a bundle`)
    }
    if (col.encryption === ENCRYPTION_DELEGATED) {
      errors.push(
        `Collection "${name}": remote collections cannot use "${col.encryption}" encryption (server cannot replicate opaque client-encrypted blobs)`,
      )
    }
    if (
      (remote.writeMode === "push_through" || remote.writeMode === "bidirectional") &&
      !remote.pushPath
    ) {
      errors.push(
        `Collection "${name}": write_mode "${remote.writeMode}" requires remote.pushPath to be set`,
      )
    }
  }

  return errors
}
